### PR TITLE
Integrate with Loops

### PR DESCRIPTION
**Pull request by rluodev**
## Summary of the problem

@zachlatta requested that we add HCB teenager users to Loops.

## Describe your changes

Whenever a user gets updated or created, we'll send their info to loops to create a new Loops `Contact`.

We'll also run a one-time job to create `Contact`s on Loops for all HCB teenager users.

They'll be added to loops default subscribed, and can opt out of receiving promo emails by unsubscribing through a loops email

image

### CODE REVIEW

File: app/services/user_service/sync_with_loops.rb
Comment (garyhtou) [inline app/services/user_service/sync_with_loops.rb]: i'd suggest doing a case insensitive check against https://github.com/hackclub/bank/blob/d5d2b851d37efc80413b34c82b498053814bae6a/app/models/stripe_cardholder.rb#L94

File: app/services/user_service/sync_with_loops.rb
Comment (sampoder) [inline app/services/user_service/sync_with_loops.rb]:
```suggestion
```

**Comment by rluodev** [comment]
Did not mean to mark as ready, still need to write job

**Comment by rluodev** [comment]
@garyhtou should we also add a consent checkbox on the account creation/settings page for subscribing to emails?

I'm thinking the wording should be something like:

`I'd like to receive updates and promotional emails from Hack Club`

If we don't have this, I think this might not be GDPR or other privacy law compliant.

**Comment by rluodev** [comment]
I'm having it set to subscribed by default; lmk if we should have it be unsubscribed by default

**Comment by zachlatta** [comment]
Yes, subscribed by default.

The opt out in settings is redundant because Loops manages subscription status and they opt out from the emails. Please cut that section from the page.

Please do not add an opt-in checkbox. Every teenager needs to be added by default to Loops. We can have a minimalist and small disclaimer (do not use the word "promotional" in it) if required.

File: app/services/user_service/sync_with_loops.rb
Comment (sampoder) [inline app/services/user_service/sync_with_loops.rb]: This should be pulled out of the `if @user.stripe_cardholder`. We should also be using `@user.last_seen_at`,

File: app/services/user_service/sync_with_loops.rb
Comment (sampoder) [reply app/services/user_service/sync_with_loops.rb]: `StripeCardholder` has `default_billing_address?` we should be using that, though maybe we need to make that method more robust.

File: app/models/user.rb
Comment (sampoder) [inline app/models/user.rb]: Are all these safe navigators necessary? I think this can be simplified to:

```suggestion
  def teenager?
    birthday&.after?(19.years.ago) || events.high_school_hackathon.any? || events.organized_by_teenagers.any?
  end
```

**Review by sampoder — COMMENTED** [review]
Couple of quick suggestions!

File: app/services/user_service/sync_with_loops.rb
Comment (rluodev) [reply app/services/user_service/sync_with_loops.rb]: so the thing about that is if there's no stripe_cardholder it'll error when I try to pull the address

**Comment by rluodev** [comment]
It's default subscribed, but only if the contact is newly added from HCB and does not previously exist, so we don't resubscribe them every time they update their profile.

**Comment by zachlatta** [comment]
Sounds good. Please re-subscribe them if they are new to HCB, but not if
they update their profile.

File: app/services/user_service/sync_with_loops.rb
Comment (sampoder) [reply app/services/user_service/sync_with_loops.rb]: I'm referring to:

```
h[:hcbLastLoginAt] = @user.user_sessions.order(created_at: :desc).last.created_at.to_time unless @user.user_sessions.order(created_at: :desc).last.nil?
```

Comment (sampoder) [reply app/services/user_service/sync_with_loops.rb]: Sorry, GitHub doesn't make that super clear

Comment (rluodev) [reply app/services/user_service/sync_with_loops.rb]: oop i see what you mean now

File: app/services/user_service/sync_with_loops.rb
Comment (zachlatta) [inline app/services/user_service/sync_with_loops.rb]: Please only override the address if the address in Loops is empty or if `addressLastUpdatedAt` is older than `@user.stripe_cardholder.updated_at`

Comment (zachlatta) [reply app/services/user_service/sync_with_loops.rb]: So basically - set the address if it's empty, or if HCB has a more recent address on file.

File: app/models/user.rb
Comment (garyhtou) [inline app/models/user.rb]:
```suggestion
    birthday&.after?(19.years.ago) || events.high_school_hackathon.or(events.organized_by_teenagers).any?
```

File: app/services/user_service/sync_with_loops.rb
Comment (garyhtou) [inline app/services/user_service/sync_with_loops.rb]: We're creating a new queue for every sync. Instead, we want to share one globally

File: app/services/user_service/sync_with_loops.rb
Comment (garyhtou) [inline app/services/user_service/sync_with_loops.rb]: See if you can configure Faraday to use 529 instead of rate limiting ourselves

File: app/services/user_service/sync_with_loops.rb
Comment (garyhtou) [inline app/services/user_service/sync_with_loops.rb]: See if you can share the connection setup since they're so similar.

https://github.com/hackclub/bank/blob/78bebb0446d17f8dbfd61a4ec19e598189520c53/app/services/api_service/v2/deliver_webhook.rb#L38-L44

**Comment by zachlatta** [comment]
Can I get an update on this, and timeline to have this live?

**Comment by garyhtou** [comment]
Hi! @zachlatta, Ruien is finishing up school this week. I'm meeting with him tomorrow to help unblock this PR. I estimate we'll have this live next week.

Let me know if you need this sooner and I can pull in other engineers, if necessary.

File: app/services/user_service/sync_with_loops.rb
Comment (garyhtou) [inline app/services/user_service/sync_with_loops.rb]:
```suggestion
      # To consider a contact as having an address, it must have a line 1, city, and country present.
      # Some international addresses don't have the concept for states or zip codes.
      @contact_details["addressLine1"].present? &&
        @contact_details["addressCity"].present? &&
        @contact_details["addressCountry"].present?
```

File: app/models/user.rb
Comment (garyhtou) [inline app/models/user.rb]:
```suggestion
    new_user = full_name_before_last_change.blank? && !onboarding?
    UserService::SyncWithLoops.new(user_id: id, new_user:).run if teenager?
```

**Review by garyhtou — COMMENTED** [review]
Last couple feedback (written during on call with @rluodev)

**Comment by rluodev** [comment]
@zachlatta the wording I've gone with on the user creation page is this statement "By signing up for HCB, you consent to receiving emails from Hack Club."

File: app/services/user_service/sync_with_loops.rb
Comment (rluodev) [inline app/services/user_service/sync_with_loops.rb]:
```suggestion
    def loops_has_address?
      # To consider a contact as having an address, it must have a line 1, city, and country present.
      # Some international addresses don't have the concept for states or zip codes.
      @contact_details["addressLine1"].present? &&
        @contact_details["addressCity"].present? &&
        @contact_details["addressCountry"].present?
    end
```

restore accidentally deleted method def

**Comment by rluodev** [comment]
Screenshot 2024-06-05 at 6 01 31 PM

(wording at bottom)

**Comment by zachlatta** [comment]
I think it probably makes more sense for the notice to be next to the email field. Is having a notice required? If not, let's please not have a notice. We will set it up so they quickly get a welcome email after signing up, which they can unsubscribe from if they don't want it.

File: app/services/user_service/sync_with_loops.rb
Comment (sampoder) [inline app/services/user_service/sync_with_loops.rb]: Let's make sure we edit creds before merging

File: app/views/users/edit.html.erb
Comment (sampoder) [inline app/views/users/edit.html.erb]: This looks a little awkward to me, +1 on Zach's comment

**Review by sampoder — DISMISSED** [review]
Good to go, thanks Ruien for all the great work on this!

File: app/views/users/edit.html.erb
Comment (garyhtou) [reply app/views/users/edit.html.erb]: i've removed it. we'll handle it in a separate PR!

**Review by garyhtou — APPROVED** [review]

**Pull request merged**
